client/notices.go

```diff
@@ -0,0 +1,189 @@
+
+	// Data are optional key=value pairs for this occurrence of the notice.
+	Data map[string]string
+}
+
+// Notify records an occurrence of a "client" notice with the specified options,
+// returning the notice ID.
+func (client *Client) Notify(opts *NotifyOptions) (string, error) {
+	var payload = struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after,omitempty"`
+		Data        map[string]string `json:"data,omitempty"`
+	}{
+		Action: "add",
+		Type:   "client",
+		Key:    opts.Key,
+		Data:   opts.Data,
+	}
+	if opts.RepeatAfter != 0 {
+		payload.RepeatAfter = opts.RepeatAfter.String()
+	}
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(&payload); err != nil {
+		return "", err
+	}
+
+	result := struct {
+		ID string `json:"id"`
+	}{}
+	_, err := client.doSync("POST", "/v1/notices", nil, nil, &body, &result)
+	if err != nil {
+		return "", err
+	}
+	return result.ID, err
+}
+
+type NoticesOptions struct {
+	// Type, if set, includes only notices of this type.
+	Type NoticeType
+	// Key, if set, includes only notices with this key.
+	Key string
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// Notice is a notification whose identity is the combination of Type and Key
+// that has occurred Occurrences number of times.
+type Notice struct {
+	ID            string            `json:"id"`
+	Type          NoticeType        `json:"type"`
+	Key           string            `json:"key"`
+	FirstOccurred time.Time         `json:"first-occurred"`
+	LastOccurred  time.Time         `json:"last-occurred"`
+	LastRepeated  time.Time         `json:"last-repeated"`
+	Occurrences   int               `json:"occurrences"`
+	LastData      map[string]string `json:"last-data,omitempty"`
+	RepeatAfter   time.Duration     `json:"repeat-after,omitempty"`
+	ExpireAfter   time.Duration     `json:"expire-after,omitempty"`
+}
+
+type NoticeType string
+
+const (
+	// Recorded whenever a change is updated: when it is first spawned or its
+	// status was updated.
+	NoticeChangeUpdate NoticeType = "change-update"
+
+	// A client notice reported via the Pebble client API or "pebble notify".
+	// The key and data fields are provided by the user. The key must be in
+	// the format "mydomain.io/mykey" to ensure well-namespaced notice keys.
+	NoticeClient NoticeType = "client"
+
+	// Warnings are a subset of notices where the key is a human-readable
+	// warning message.
+	NoticeWarning NoticeType = "warning"
+)
+
+type jsonNotice struct {
+	Notice
+	RepeatAfter string `json:"repeat-after,omitempty"`
+	ExpireAfter string `json:"expire-after,omitempty"`
+}
+
+// Notice fetches a single notice by ID.
+func (client *Client) Notice(id string) (*Notice, error) {
+	var jn *jsonNotice
+	_, err := client.doSync("GET", "/v1/notices/"+id, nil, nil, nil, &jn)
```

Comment:
probably out of scope for this PR, but I think this is a great example where an args struct would be appropriate. That way, we can name the arguments (what are all these nils?), and not have to explicitly pass in nil values

Reply:
Yeah, agreed: both that this is a good idea, and that it's out of scope for this PR. It's not really causing pain, but I'd happily approve a PR that made this change.

Incidentally, this was proposed as part of [the PR to expose client.DoSync etc](https://github.com/canonical/pebble/pull/289/files), but we decided not to go ahead with that one as it was considered an implementation detail of the client. But I still think it'd be fine to make the internal `doX` funcs take args structs.

Reply:
We already have that struct now, after the Requester PR has landed.